feature: implement EngineData parsing

scoiatael opened the pull request:
fixes https://github.com/webtoon/psd/issues/6

This feature should be more or less finished - during the code review I'll run it on sample designs to verify parsing works consistently.

pastelmind reviewed (commented):
First, thank you for working on the PR.

I'll be honest: [I have been building an EngineData parser myself](https://github.com/pastelmind/EngineData). I began working on it around the 27th of July, a couple of days after you volunteered to write the PR.

Please understand that I had no intention of undermining your hard work. I had been planning to attack the problem myself for a long time, and your initial suggestion motivated me to make one last attempt. I apologize for not telling you this sooner. I didn't want to stop you in case my solution failed, and I did not anticipate that you would submit a PR in such a short time.

## Performance

I ran some quick benchmarks on my code as well as yours (with slight modifications), and the results currently favor mine by 5-60% on different browsers:

- Chrome 104.0
<img width="300" alt="chrome 104 0" src="https://user-images.githubusercontent.com/29533079/182444946-977fe1cb-d626-49d1-9dcb-10dbe04409ec.png">

- Firefox 103.0.1
<img width="290" alt="firefox 103 0 1" src="https://user-images.githubusercontent.com/29533079/182444972-48833043-487c-4fc6-9bc9-c96f5b4cd60a.png">

- Safari 15.6
<img width="289" alt="safari 15 6" src="https://user-images.githubusercontent.com/29533079/182444990-7f0f7aab-bdc4-4110-9066-18980a2a155b.png">

(My benchmarks were done on a MacBook Pro 13-inch / M1 / 2020)

## Integration

My parser is almost done and only needs to be integrated into the rest of the library. If you wish to work on your parser's performance, I'm ready to wait for further submissions. Otherwise, we could reuse your code (e.g. `Layer.ts`) to integrate my parser into `@webtoon/psd`.

scoiatael commented on packages/psd/src/engineData/parser.ts:
```diff
@@ -0,0 +1,106 @@
+      it = this.advance();
+    }
+    switch (it.type) {
+      case TokenType.Name:
```
Are you sure about that interpretation?

At least in PDF spec `/`-prefixed sequences are called "Name Objects" and are
> an atomic symbol uniquely defined by a sequence of any characters
(8-bit values) except null (character code 0). Uniquely defined means that any two name objects made up of
the same sequence of characters denote the same object.

A dictionary on the other hand:
> is an associative table containing pairs of objects, known as the dictionary’s entries. The first
element of each entry is the key and the second element is the value. The key shall be a name (unlike
dictionary keys in PostScript, which may be objects of any type). The value may be any kind of object, including
another dictionary

And so names are in no way tied to dictionaries - and as such sequence `<< /abc /abc >>` should mean "a dictionary containing one object, under key of Name "abc" and having value of Name "abc" - and both key and value referencing the same object (although iirc not many implementations care about that 2nd part). There's even an example of dict with values being names in the PDF spec:
```
EXAMPLE << /Type /Example
/Subtype /DictionaryExample
/Version 0 . 01
/IntegerItem 12
/StringItem ( a string )
/Subdictionary << /Item1 0 . 4
/Item2 true
/LastItem ( not ! )
/VeryLastItem ( OK )
>>
>>
```

See [PDF spec](https://web.archive.org/web/20220226063926/https://www.adobe.com/content/dam/acom/en/devnet/pdf/pdfs/PDF32000_2008.pdf) - section 7.3.5 (Name objects) and 7.3.7 (Dictionary objects).

pastelmind replied:
That is interesting. I did not know that EngineData is based on the PDF format.

Have you observed any PSD file that actually uses `/name` as anything other than a dictionary key? Personally I have not.

scoiatael replied:
I doubt there's any official mention anyplace - but seeing how both are Adobe-based and how Adobe likes to push PDF in places where they have to put unstructured data it'd IMO make sense. That's why I based lexer and parser on PDF spec :)

scoiatael commented on packages/psd/src/engineData/lexer.ts:
```diff
@@ -0,0 +1,217 @@
+        }
+        textParts.push(decoder.decode(raw));
+        readAhead.pass(1); // skip over \\
+        textParts.push(String.fromCharCode(readAhead.take(1)[0])); // un-escape character
```
Thanks! Fixed accordingly - as a bonus the code simplified a little :)

scoiatael commented:
No hard feelings, but next time you could mention you are willing to give a feature one more try - I could have implemented something else in the same time and library would've benefited from that :)

---
As for this PR - I implemented fix to the CJK parsing bug and added test for it. I also migrated parser to stack-based - to make optimizations a little easier (now flamegraphs can properly merge similar invocations because stack looks largely the same).

That said, I couldn't find any meaningful optimizations (more that ~5% on Node.js). I suspect your hand-written lexer simply behaves better that what is generated from the `async`-function in my case :)

---
As for next steps - I'm literally indifferent as to which implementation ends up on master - as long as any does :) As you are the maintainer, the choice is yours (and your colleagues). The code in this PR is also yours to use as you see fit - e.g. for integrating your library if that speeds up development and such is the choice. Just let me know ;)

---
One final question: would it make sense to re-write this part in Rust? Since it looks like parsing EngineData might take a large portion of total design parsing time.

pastelmind commented:
Thank you. I handled the situation poorly, and will not make the same mistake again.

I doubt parsing EngineData would be appropriate for Rust conversion. Decoding layer images is usually the most time- and memory-intensive task by far (hundreds of milliseconds per layer). Parsing an EngineData document with JS takes < 1ms when optimized, and I doubt Rust+WebAssembly would have a noticeable impact.

scoiatael commented:
Ran some benchmarks; looks like for overall file opening process your parser provides a significant speedup:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `cat samples.txt \| xargs node parse.mjs scoiatael` | 6.868 ± 0.086 | 6.727 | 7.014 | 1.14 ± 0.05 |
| `cat samples.txt \| xargs node parse.mjs pastelmind` | 6.049 ± 0.262 | 5.682 | 6.361 | 1.00 |

(where `parse.mjs` is [here](https://gist.github.com/scoiatael/5c28288df5cb4471812bdfde765f7538), Node is `v18.2.0` and hardware is `CPU: 11th Gen Intel i7-1185G7 (8) @ 4.800GHz`)

@pastelmind Maybe the simplest way forward it to just merge this PR and then replace my implementation with yours?

scoiatael commented:
@pastelmind had some spare time and managed to find pretty substantial optimizations; right now the results are:

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `cat samples.txt \| xargs node parse.mjs scoiatael` | 4.793 ± 0.076 | 4.677 | 4.899 | 1.00 |
| `cat samples.txt \| xargs node parse.mjs pastelmind` | 5.889 ± 0.052 | 5.815 | 5.977 | 1.23 ± 0.02 |

pastelmind reviewed (approved):
I was sick for most of the week and only got around to reviewing your PR just now.

- Awesome work on optimization! It seems that avoiding unnecessary memory allocations greatly boosted your lexer's performance.
- Could you clarify <q>memoize results for each character instead of lookup</q> in https://github.com/webtoon/psd/pull/43/commits/cd40e0848c8ca76c072c3b46b886bf719a6e9568<wbr>? I don't see any memoizing code.
- `Cursor` is becoming somewhat burdened. We could split it into two variants--the regular `Cursor` for parsing standard structures, and a specialized `EngineDataCursor` exclusively for the lexer. This isn't needed now, though.

I will accept the PR for now.

scoiatael commented:
> memoize results for each character instead of lookup

refers to [`STRING_TOKEN_JT`](https://github.com/webtoon/psd/commit/cd40e0848c8ca76c072c3b46b886bf719a6e9568#diff-87cdd246cfb9a56f2fce89ee3b898ce5cfa34936bedafee1e63c5ef85ab108caR102-R106) - which pre-computes a boolean value for each character instead of doing two map lookups at runtime :) It was originally intended to be a specialized jump-table (hence JT suffix) but turned out that simple boolean array works best for this case.

> It seems that avoiding unnecessary memory allocations greatly boosted your lexer's performance.

If you are referring to using Array instead of Generator then not really - I did it just to have clean demarkation line when benchmarking with 0x. This change can be easily reverted if we ever start to run into memory issues :)
